File: packages/sdk-go/conformance/json_decoding_test.go
package conformance

import (
	"encoding/json"
	"strconv"
	"strings"
	"testing"

	"github.com/gemini/developer-platform/packages/sdk-go/types"
)


func TestJSONDecodingConformance(t *testing.T) {
	manifest := LoadManifest(t)
	found := false
	for _, suite := range manifest.Suites {
		if suite.Kind != "jsonDecoding" {
			continue
		}
		found = true
		for _, caseID := range suite.Cases {
			t.Run(caseID, func(t *testing.T) {
				var fixture FixtureJSON
				if err := json.Unmarshal(LoadCase(t, suite.ID, caseID), &fixture); err != nil {
					t.Fatalf("decode fixture: %v", err)
				}
				if fixture.Kind != "jsonDecoding" {
					t.Fatalf("fixture kind = %q, want jsonDecoding", fixture.Kind)
				}
				if fixture.Field != "v" {
					t.Fatalf("unsupported fixture field %q", fixture.Field)
				}

				decoder := json.NewDecoder(strings.NewReader(fixture.Raw))
				decoder.UseNumber()
				switch fixture.Expect.ValueKind {
				case "integer":
					var value struct {
						V int64 `json:"v"`
					}
					if err := decoder.Decode(&value); err != nil {
						t.Fatalf("decode integer: %v", err)
					}
					if got := strconv.FormatInt(value.V, 10); got != fixture.Expect.Text {
						t.Fatalf("decoded integer = %q, want %q", got, fixture.Expect.Text)
					}
				case "decimal":
					var value struct {
						V types.Decimal `json:"v"`
					}
					if err := decoder.Decode(&value); err != nil {
						t.Fatalf("decode decimal: %v", err)
					}
					if got := value.V.String(); got != fixture.Expect.Text {
						t.Fatalf("decoded decimal = %q, want %q", got, fixture.Expect.Text)
					}
				default:
					t.Fatalf("unsupported value kind %q", fixture.Expect.ValueKind)
				}
			})
		}
	}
	if !found {
		t.Fatal("manifest has no jsonDecoding suite")
	}
}

File: packages/sdk-typescript/src/tests/conformance/json-decoding.test.ts
import test from "node:test";
import assert from "node:assert/strict";

import { parseLosslessJson } from "../../transport/http.js";
import { isBoundaryObject } from "../../utils/boundary-value.js";
import { loadCase, loadManifest } from "./support/fixtures.js";

test("jsonDecoding fixtures preserve wire numeric semantics", async () => {
  const manifest = await loadManifest();
  const suite = manifest.suites.find((candidate) => candidate.kind === "jsonDecoding");
  assert.ok(suite, "manifest must declare a jsonDecoding suite");

  for (const caseId of suite.cases) {
    await test(caseId, async () => {
      const fixture = await loadCase(suite.id, caseId);
      const raw = fixture.raw;
      const field = fixture.field;
      const expected = fixture.expect;
      assert.equal(typeof raw, "string");
      assert.equal(typeof field, "string");
      assert.ok(isBoundaryObject(expected));
      const parsed = parseLosslessJson(raw as string);
      assert.ok(isBoundaryObject(parsed), "json fixture must decode to an object");
      const value = parsed[field as string];
      assert.notEqual(value, undefined, `fixture field ${String(field)} must be present`);

      switch (expected.valueKind) {
        case "integer":
          assert.equal(String(value), String(expected.text));
          if (BigInt(String(expected.text)) > BigInt(Number.MAX_SAFE_INTEGER) ||
            BigInt(String(expected.text)) < BigInt(Number.MIN_SAFE_INTEGER)) {
            assert.equal(typeof value, "bigint", "unsafe integers must be promoted to bigint");
          } else {
            assert.equal(typeof value, "number", "safe integers must remain numbers");
          }
          break;
        case "decimal":
          if (typeof value === "string") {
            assert.equal(value, expected.text, "decimal strings must remain exact strings");
          } else {
            assert.equal(typeof value, "number", "decimal numbers must remain numbers");
            assert.equal(value, Number(expected.text), "decimal number value must be preserved");
          }
          break;
        default:
          assert.fail(`unsupported json fixture valueKind: ${String(expected.valueKind)}`);
      }
    });
  }
});
